### ISSUE TITLE
Allow tune_grid() like function to finalize_model() with best metric

### ISSUE THREAD

**Issue by nabsiddiqui** [tidymodels/tune#639]
Currently, workflow requires that you create an intermediary object to select the best metric and update a workflow with it. For instance:

```r
grid_result <- some_workflow %>%
  tune_grid(resamples = some_resamples, grid = some_grid) %>%
  select_best()

some_workflow() %>%
finalize_workflow(grid_result)
```

It would be easier to have a tune grid option that automatically updated the workflow itself with the best metric since many are probably going to use the best metric to do this. It could take the name of the metric as an optional second argument. 

```r
some_workflow %>%
   tune_workflow(resamples = some_resamples, grid = some_grid)
```

**Comment by simonpcouch** [comment]
Hey there @nabsiddiqui! Thanks for the issue.

We introduced a helper in #586 called `fit_best()` that is somewhat similar to what you proposed, but additionally _fits_ the workflow with the best hyperparameter combination.

``` r
library(tidymodels)

tune_res <- 
  tune_grid(
    nearest_neighbor("classification", neighbors = tune()), 
    Class ~ ., 
    vfold_cv(two_class_dat),
    control = control_grid(save_workflow = TRUE)
  )

# note best roc_auc is with `neighbors = 15`
collect_metrics(tune_res)
#> # A tibble: 16 × 7
#>    neighbors .metric  .estimator  mean     n std_err .config             
#>        <int> <chr>    <chr>      <dbl> <int>   <dbl> <chr>               
#>  1         2 accuracy binary     0.726    10  0.0168 Preprocessor1_Model1
#>  2         2 roc_auc  binary     0.792    10  0.0175 Preprocessor1_Model1
#>  3         3 accuracy binary     0.726    10  0.0168 Preprocessor1_Model2
#>  4         3 roc_auc  binary     0.814    10  0.0165 Preprocessor1_Model2
#>  5         5 accuracy binary     0.789    10  0.0167 Preprocessor1_Model3
#>  6         5 roc_auc  binary     0.845    10  0.0147 Preprocessor1_Model3
#>  7         7 accuracy binary     0.800    10  0.0152 Preprocessor1_Model4
#>  8         7 roc_auc  binary     0.856    10  0.0136 Preprocessor1_Model4
#>  9         9 accuracy binary     0.801    10  0.0126 Preprocessor1_Model5
#> 10         9 roc_auc  binary     0.862    10  0.0137 Preprocessor1_Model5
#> 11        10 accuracy binary     0.805    10  0.0141 Preprocessor1_Model6
#> 12        10 roc_auc  binary     0.863    10  0.0139 Preprocessor1_Model6
#> 13        12 accuracy binary     0.800    10  0.0130 Preprocessor1_Model7
#> 14        12 roc_auc  binary     0.867    10  0.0134 Preprocessor1_Model7
#> 15        15 accuracy binary     0.809    10  0.0119 Preprocessor1_Model8
#> 16        15 roc_auc  binary     0.871    10  0.0124 Preprocessor1_Model8

# a shorthand for `select_best() %>% finalize_workflow() %>% fit()`
# note `k = 15`
fit_best(tune_res)
#> ══ Workflow [trained] ══════════════════════════════════════════════════════════
#> Preprocessor: Formula
#> Model: nearest_neighbor()
#> 
#> ── Preprocessor ────────────────────────────────────────────────────────────────
#> Class ~ .
#> 
#> ── Model ───────────────────────────────────────────────────────────────────────
#> 
#> Call:
#> kknn::train.kknn(formula = ..y ~ ., data = data, ks = min_rows(15L,     data, 5))
#> 
#> Type of response variable: nominal
#> Minimal misclassification: 0.1871049
#> Best kernel: optimal
#> Best k: 15
```

<sup>Created on 2023-03-15 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

`?fit_best()` has more information on usage. Note that these changes are not yet on CRAN, so you will need to install the dev version with `pak::pkg_install("tidymodels/tune")`.

**Comment by nabsiddiqui** [comment]
Thank you @simonpcouch. I think this works well. If it is truly just a shortcut though, I think it makes more sense to do last_fit() rather than just fit(). 

Presumably, after getting the metric, best practices would entail checking the results on the testing set. For me, last_fit is very useful because you can then check the metrics and extract the model easily using the extract functions: https://workflows.tidymodels.org/reference/extract-workflow.html

You can add an additional last_fit() but this has the problem of training twice, which is fine for small datasets but may be problematic with large ones. The example below is on the quantifying art history dataset from Tidy Tuesday a while ago:

```r
artists_final_fit <- workflow() %>%
    add_recipe(artists_recipe)%>%
    add_model(artists_model) %>%
    tune_grid(resamples = artists_folds, 
              grid = tree_grid, 
              control=control_grid(save_workflow=TRUE)) %>%
    fit_best() %>% 
    last_fit(split=artists_split)
```

**Comment by simonpcouch** [comment]
Yup, heard that we want to avoid having to train on the full training set twice.

The interface that you suggest, then, feels well-supported by `last_fit()`. I'm not sure that any "intermediate objects" are needed in that case:

``` r
library(tidymodels)

wf <- workflow(Class ~ ., decision_tree("classification", min_n = tune()))

splits <- initial_split(two_class_dat)
folds <- vfold_cv(training(splits))

res <- 
  tune_grid(wf, folds) %>%
  select_best("roc_auc") %>%
  finalize_workflow(wf, .) %>%
  last_fit(splits)

res
#> # Resampling results
#> # Manual resampling 
#> # A tibble: 1 × 6
#>   splits            id               .metrics .notes   .predictions .workflow 
#>   <list>            <chr>            <list>   <list>   <list>       <list>    
#> 1 <split [593/198]> train/test split <tibble> <tibble> <tibble>     <workflow>
```

<sup>Created on 2023-03-15 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by nabsiddiqui** [comment]
I feel I'm missing some basic R piping that will allow it to all be one simple command. I have the following:

```r
wf <- workflow() %>%
  add_model(artists_model) %>%
  add_recipe(artists_recipe) 

wf <- wf() %>% 
  tune_grid(resamples = artists_folds, grid = tree_grid) %>%
  select_best() %>% 
  finalize_workflow(wf, .) %>%
  last_fit(artists_split)
```

But, I am looking for the following. Somewhere I'm messing up the call to `finalize_workflow()`. Apologize for the question which likely has a simple answer.

```r
wf <- workflow() %>%
  add_model(artists_model) %>%
  add_recipe(artists_recipe) %>% 
  tune_grid(resamples = artists_folds, grid = tree_grid) %>%
  select_best() %>% 
  finalize_workflow() %>%
  last_fit(artists_split)
```

**Comment by simonpcouch** [comment]
Ah, I see. This looks right—the workflow specification is indeed an object that would need to be assigned to in this pipeline. The specification of workflow feels to me like a reasonably important result to require a persistent object, and the unfitted workflow doesn't take up much memory comparatively, so I don't imagine we'll make any changes here.

If not having that `wf` object around is important to you, you could inline the workflow specification:

``` r
wf <- workflow(artists_recipe, artists_model) %>%
  tune_grid(resamples = artists_folds, grid = tree_grid) %>%
  select_best() %>% 
  finalize_workflow(workflow(artists_recipe, artists_model), .) %>%
  last_fit(artists_split)
```

**Comment by nabsiddiqui** [comment]
Thank you. I think this works for my purposes.
